Add CLI for converting v2 metadata to v3

K-Meech opened the pull request:
For https://github.com/zarr-developers/zarr-python/issues/1798

Adds a CLI using [`typer`](https://typer.tiangolo.com/) to convert v2 metadata (`.zarray` / `.zattrs`...) to v3 metadata `zarr.json`. 

To test, you will need to install the new optional cli dependency e.g.
`pip install -e ".[remote,cli]"`

This should make the `zarr-converter` command available e.g. try:
```
zarr-converter --help
zarr-converter convert --help
zarr-converter clear --help
```
`convert` adds `zarr.json` files to every group / array, leaving the v2 metadata as-is. A zarr with both sets of metadata can still be opened with `zarr.open`, but will give a UserWarning: `Both zarr.json (Zarr format 3) and .zarray (Zarr format 2) metadata objects exist... Zarr v3 will be used.`. This can be avoided by passing `zarr_format=3` to `zarr.open`, or by using the `clear` command to remove the v2 metadata.

`clear` can also remove v3 metadata. This is useful if the conversion fails part way through e.g. if one of the arrays uses a codec with no v3 equivalent.

All code for the cli is in `src/zarr/core/metadata/converter/cli.py`, with the actual conversion functions in `src/zarr/core/metadata/converter/converter_v2_v3.py`. These functions can be called directly, for those who don't want to use the CLI (although currently they are part of `/core` which is considered private API, so it may be best to move them elsewhere in the package).

Some points to consider:
- I had to modify `set_path` from `test_dtype_registry.py` and `test_codec_entrypoints.py`, as they were causing the CLI tests to fail if they were run after. This seems to be due to the `lazy_load_list` of the numcodecs codecs registries being cleared, meaning they were no longer available in my code which finds the `numcodecs.zarr3` equivalent of a numcodecs codec.
- I tested this on local zarr images, so it would be great if someone with access to s3 / google cloud etc., could try it out on some small example images there.
- I'm happy to add docs about how to use the CLI, but wanted to get feedback on the general structure first

TODO:
* [X] Add unit tests and/or doctests in docstrings
* [X] Add docstrings and API docs for any new/modified user-facing classes and functions
* [x] New/modified features documented in `docs/user-guide/*.rst`
* [x] Changes documented as a new file in `changes/`
* [x] GitHub Actions have all passed
* [x] Test coverage is 100% (Codecov passes)

d-v-b commented:
this is awesome! A few high-level suggestions:

- can we have the name of the cli be just `zarr`, where `convert` is a subcommand: 
  ```
  $ zarr --help
   # prints info including available commands
  $ zarr convert --help
  # info about the convert operation
  ```

- can the 2->3 migration just be a particular invocation of a more general command, like
  ```zarr convert source/data.zarr --in-place --zarr-format 3```

  I'm not sure how to express the "clean up a broken conversion" process through this API, but it should be doable

- is "convert" the best name here? "resave" also seems good. I'm not opposed to "convert", just interested in exploring the space of names a bit.

d-v-b commented:
> I think the only situation this doesn't cover is if the conversion fails part way through + you just want to remove any zarr.json that were written. Perhaps it would be best to prevent this scenario from occurring in the first place? I could always run convert_v2_to_v3 inside a try / except, then run remove_metadata if any exceptions are encountered. This could be optional behaviour with something like --clear

Attempting to clean up on failure is a good idea, but that can also fail. It would still be possible for the program to be interrupted while `remove_metadata` is being run, so I don't think there's anything we can do prevent conversion from failing. But as long as we are only creating json documents, and we aren't deleting the originals (the v2 metadata), then re-running should be cheap.  This seems like a detail that we can iron out later on.

> --in-place would create v3 metadata + remove any v2 metadata (without this specified, both v2/v3 metadata would remain)

There might actually be situations where people want the v2 and v3 metadata in the same hierarchy, and this is not against the rules of the spec. So I'd say that creating v3 metadata should be a separate (and precedent) operation from deleting the v2 metadata. `--in-place` should be considered shorthand for this:

```
$ zarr convert foo.zarr/bar foo.zar/bar
```
i.e., taking a v2 hierarchy from foo.zarr/bar, and creating a v3 hierarchy at foo.zarr/bar. Thus I think 
```
$ zarr convert foo.zarr baz.zarr
```
i.e., resaving the hierarchy somewhere else, should also be supported. For now users would be responsible for copying chunks in this case. 

> --allow-overwrite would remove any existing v3 metadata, then create new (otherwise it errors if an existing zarr.json is encountered)

that seems right, but instead of `--allow-overwrite` I would just do `--overwrite`.

And a broader request: I am sure that while working on this kind of low-level tool, you find tasks that should be direct, but are made difficult by design choices in zarr-python, or the lack of low-level features. If you have the patience for it, please keep track of the stuff we could change in the core library to make this kind of tool easier to write, and feel free to open issues to track these features.

K-Meech commented:
Ok - that all sounds good. How about the following for an updated CLI:

The base command is `zarr convert path/to/input.zarr` - this creates v3 metadata, writing it in-place back to `input.zarr`, leaving the v2 metadata as-is. There are then the following optional commands on top:
- `--output path/to/output.zarr` - if provided, v3 metadata will be written to this separate output location (leaving `input.zarr` completely unchanged). Only the metadata will be created (no copying of chunks).
- `--overwrite`: allows overwrite of v3 metadata at the output location (either `path/to/input.zarr` or `path/to/output.zarr` if provided)
- `--remove-v2`: removes any v2 metadata at the output location

I think it's probably worth also keeping the `clear` command to give maximum flexibility. I've certainly used it a lot during testing of the CLI e.g. when converting in-place, then doing some testing of the v3 output to check it looks ok, then later removing the v2 metadata when it's no longer needed.

d-v-b commented:
> * `--output path/to/output.zarr` - if provided, v3 metadata will be written to this separate output location (leaving `input.zarr` completely unchanged). Only the metadata will be created (no copying of chunks).

This design would make the in-place conversion the default, and writing the hierarchy to a new path would be an opt-in kind of thing. Speaking for myself, I think I would expect the convert-to-somewhere-else mode to be the default, and the in-place conversion to be opt-in. It might be helpful to sample opinions from other folks here (cc @zarr-developers/python-core-devs). In any case this is a minor thing.


> I think it's probably worth also keeping the `clear` command to give maximum flexibility. I've certainly used it a lot during testing of the CLI e.g. when converting in-place, then doing some testing of the v3 output to check it looks ok, then later removing the v2 metadata when it's no longer needed.

That makes perfect sense. Since "clear" is kind of ambiguous on its own, and since the `convert` command is pretty narrowly scoped to converting from v2 to v3 data, I wonder if we should rename this command to something like `convert-v3`, `migrate-v3`, etc, something that makes it really obvious that all the options will be scoped to the process of moving from v2 data to v3 data. This would free up the `convert` command for later use, e.g. things like changing chunking, compression, etc.

d-v-b commented:
also, in case anyone wants to try this tool out, here's a 1-liner with uvx:

```
uvx --with typer --from git+https://github.com/K-Meech/zarr-python/@km/v2-v3-conversion zarr-converter --help
```

dstansby commented on src/zarr/core/metadata/converter/converter_v2_v3.py:
```diff
@@ -0,0 +1,256 @@
+        sync(_save_v3_metadata(zarr_v2, array_metadata_v3, dry_run=dry_run))
+
+
+async def remove_metadata(
```
Am I right in thinking this isn't required by the conversion functionality? If not, I'd be tempted not to implement it at least in this PR to keep this PR tightly scoped and therefore easier to review. I have some comments/questions about performance implications here, but thought maybe that could be discussed in a follow up PR?

dstansby replied:
Sorry, I just read the comment thread, feel free to ignore this and leave clear here!

dstansby commented on src/zarr/core/metadata/converter/converter_v2_v3.py:
```diff
@@ -0,0 +1,256 @@
+        fill_value=metadata_v2.fill_value,
+        codecs=codecs,
+        attributes=metadata_v2.attributes,
+        dimension_names=None,
```
A thought - it might be nice to give the option of adding dimension names during conversion? That could go in a follow up PR though.

dstansby commented on src/zarr/core/metadata/converter/converter_v2_v3.py:
```diff
@@ -0,0 +1,256 @@
+    filters_codecs = [_find_numcodecs_zarr3(filter) for filter in metadata_v2.filters]
+    for codec in filters_codecs:
+        if not isinstance(codec, ArrayArrayCodec):
+            raise TypeError(f"Filter {type(codec)} is not an ArrayArrayCodec")
```
If filters aren't ArrayArray codecs, instead of erroring here would it be possible to include them somehow in the new metadata?

d-v-b replied:
for more background, in the zarr v2 format, all the elements of `filters`, and the value assigned to `compressor`, are the same kind of thing, from the metadata POV. 

the zarr v3 format introduced the idea of an "array array codec", "array bytes codec", "bytes bytes codec", but there was no effort to guarantee that all v2 metadata documents could be easily translated into the v3 form.

d-v-b replied:
> I can change this to pool the filters / compressors together and handle them in the same way. I guess the order is still important though? Should it still be: TransposeCodec > filter codecs > BytesCodec > compressor codec? 

This makes sense, and yes the order is still important, but what you have looks good.

An important case to consider is arrays that use vlen-utf8 or vlen-bytes codecs for compressing variable-length strings or variable-length bytes. For these arrays, we don't need a `BytesCodec`, because the vlen-utf8 and vlen-bytes codecs do the same job as the `bytescodec` (i.e., they are effectively `ArrayBytesCodecs`). But I think we could add support for those arrays later, if you want to ensure that this PR gets done sooner.

d-v-b replied:
yet another situation where it would be great if we had a big collection of various zarr arrays on github, each using different compression routines, for testing this kind of thing

K-Meech replied:
Thanks @d-v-b . Ok - I'll keep the order the same, but remove the exceptions for `ArrayArrayCodec` / `BytesBytesCodec` for filters and compressors (allowing them to be of any type). I'll also process them both in the same way, so they're treated as identical.

For the `vlen` codecs, I think the current code will error saying that there is no `numcodecs.zarr3` equivalent. My preference would probably be to add support for these in a future PR - as this is already quite large!

K-Meech replied:
@d-v-b I've made these changes locally, but the order of codecs seems to cause issues. 

For example, I created a v2 array with: 
- `"filters": [{"id": "zstd", "level": 3}], "compressor": None"`
Then converted it to v3: 
- `codecs` are `ZstdCodec` followed by `BytesCodec`. 

When I try to open the v3 array I get: `TypeError: Invalid codec order. ArrayBytes codec BytesCodec(endian=<Endian.little: 'little'>) must be preceded by an ArrayArrayCodec. Got <class 'zarr.codecs.zstd.ZstdCodec'> instead.` 

Opening the v2 array directly works - so I guess it is doing the equivalent of `BytesCodec` first then `ZstdCodec` after (rather than enforcing filters > BytesCodec > compressors). Should I dynamically put the `BytesCodec` in a different position based on the types of filters and compressors?

K-Meech replied:
Ok thanks - I'm away for the next week, but I'll give this another go when I'm back.
I'll convert all of the filters and compressors, putting them into a list in order (filters followed by compressors). At this stage they'll all have types of `ArrayArrayCodec`, `ArrayBytesCodec`, or `BytesBytesCodec`. Then I can find the first `BytesBytesCodec` and place the `BytesCodec` in the slot before this.

Seems like there may be a lot of edge cases here though - e.g. [`numcodecs.zarr3`](https://numcodecs.readthedocs.io/en/stable/zarr3.html#serializers-array-to-bytes-codecs) has some `ArrayBytesCodec` listed - so in this case I guess we wouldn't need to insert the `BytesCodec` at all. I'm curious how this works with opening an array with v2 metadata directly - is there a function somewhere in the codebase that does this logic for figuring out when/if this should happen in the filters/compressors sequence?

dstansby reviewed (commented):
I have left some comments - not finished reviewing yet, but I have to bounce for a few hours now and thought having some to think about/work on might be helpful in the meantime. Will carry on reviewing later today!

K-Meech commented:
Thanks for the comments @dstansby! I'll start making the changes you suggested in the review above.

As for the overall structure - how about:

`zarr migrate v3` command:
- Simplest call is either `zarr migrate v3 path/to/input.zarr path/to/output.zarr` (add metadata at separate location) or `zarr migrate v3 path/to/input.zarr` (in place)
- The following extra options are available:
  - `--overwrite`: allows overwrite of existing v3 metadata at the output location
  -  `--remove-v2`: removes any v2 metadata at the output location after conversion
  -  `--dry-run`: log changes that would be made, without making them
  - Previously I'd supported a `path` option which, similar to [`zarr.open`](https://zarr.readthedocs.io/en/stable/api/zarr/index.html#zarr.open), specified the path within the store to open. Now that there's possibly separate input / output locations - this could be expanded to `--input-path` and `--output-path`.
  - The underlying functions e.g. [convert_v2_v3](https://github.com/K-Meech/zarr-python/blob/km/v2-v3-conversion/src/zarr/core/metadata/converter/converter_v2_v3.py#L37) also support `storage_options` similar to [`zarr.open`](https://zarr.readthedocs.io/en/stable/api/zarr/index.html#zarr.open) - should this be supported in the CLI too?

`zarr remove-metadata` command:
- Simplest call is `zarr remove-metadata v3 /path/to/.zarr` or `zarr remove-metadata v2 /path/to/.zarr`
- Options are:
  - `--path`: path within the store to open
  - `--dry-run`: log files that would be deleted, without removing them
  - `--force`: allows deletion of metadata, when there is no valid alternative e.g. can delete v3 metadata, when no valid v2 metadata OR delete v2 metadata when there is no valid v3 metadata

K-Meech commented:
- `--overwrite` is mainly provided for the situation where a conversion fails part way through, and you want to run it again. For example, say you were converting a large remote zarr and your connection failed part way through. You may have `zarr.json` in some locations in your hierarchy, but not others, which would need to overwritten by a new conversion call. (You could also call `remove-metadata` first, then `migrate`, but this is a shortcut for that)

- Happy to change to `--remove-v2-metadata` to make this clearer.

- Ok - I can remove the `path` option, and people will need to provide the full path to the array/group they want to convert in the input/output store part.

- Happy to keep `storage_options` for a future PR, as it isn't super-simple to allow an arbitrary Python dict of options

K-Meech commented:
I've updated the structure of the CLI - hopefully this addresses both @dstansby and @d-v-b 's comments! You should be able to test with:
```
zarr --help
zarr migrate --help
zarr remove-metadata --help
```
I haven't addressed [this comment yet](https://github.com/zarr-developers/zarr-python/pull/3257#discussion_r2221835719), but will do so in the next round of changes (I know @dstansby had some additional comments to make on the converter implementation).

One known issue:
- If you run `zarr migrate --dry-run` with a non-existent location e.g. `zarr migrate v3 data/images/example-1.zarr --dry-run`, it will create an empty directory at that location. This is related to `r+` opens - reported on this issue: https://github.com/zarr-developers/zarr-python/issues/3295

dstansby commented:
Very nice! I had a play with it locally and it worked well. I'll do a fuller review of the code now.

dstansby commented on src/zarr/core/metadata/converter/migrate_to_v3.py:
```diff
@@ -0,0 +1,303 @@
+    )
+
+
+def _convert_filters(metadata_v2: ArrayV2Metadata) -> list[ArrayArrayCodec]:
```
I think it would make slightly more sense to pass filters, instead of the whole metadata here.
```suggestion
def _convert_filters(filters) -> list[ArrayArrayCodec]:
```
(I probably got the typing wrong)

dstansby commented on src/zarr/core/metadata/converter/migrate_to_v3.py:
```diff
@@ -0,0 +1,303 @@
+    return cast(list[ArrayArrayCodec], filters_codecs)
+
+
+def _convert_compressor(metadata_v2: ArrayV2Metadata) -> BytesBytesCodec | None:
```
```suggestion
def _convert_compressor(compressor, dtype) -> BytesBytesCodec | None:
```
Again, I think makes for a clearer API if you explicitly just pass the compressor and type instead of the whole metadata.

dstansby commented on src/zarr/core/metadata/converter/migrate_to_v3.py:
```diff
@@ -0,0 +1,303 @@
+    if metadata_v2.compressor is None:
+        return None
+
+    compressor_name = metadata_v2.compressor.codec_id
+
+    match compressor_name:
```
```suggestion
    match metadata_v2.compressor.codec_id:
```
Since you don't use the variable again

emmanuelmathot commented:
Just FYI, I made this [converter](https://github.com/EOPF-Explorer/data-model) for ESA EOPF datasets (Zarr v2) to EOPF GeoZarr (V3). I adopted a recursive approach to copy and keep the datatree as well as the consolidated metadata.

K-Meech commented:
@TomNicholas - this PR will add both a CLI + functions to convert v2 to v3 metadata. Everything possible with the CLI is also available via direct function calls. 

Although, at the moment these are added to `src/zarr/core/metadata` which I don't think is part of the public API - @d-v-b / @dstansby, any suggestions for where these could be moved to make them easy to use by downstream scripts / packages?

TomNicholas commented:
@K-Meech thanks. That's actually okay for my use case - I'm already importing other private metadata-related internals. But I agree a later PR to make these public would be great.

dstansby commented:
I'd suggest putting the metadata converter python API in a new `zarr.metadata` submodule (which can contain other bits of metadata migrated from `zarr.core.metadata` later), and hiding the the CLI wrapper in a new private `zarr._cli` submodule (or `zarr._cli.py` file)

K-Meech commented on src/zarr/core/metadata/converter/cli.py:
```diff
@@ -0,0 +1,180 @@
+    else:
+        lvl = logging.WARNING
+    fmt = "%(message)s"
+    logging.basicConfig(level=lvl, format=fmt)
```
I could do something like:
```
if verbose:
        logger.setLevel(logging.INFO)
    else:
        logger.setLevel(logging.WARNING)
logger.addHandler(logging.StreamHandler())
```
but the issue is this will only affect logs directly from the `cli.py` file. When using `--dry-run`, I also want to see the log of created / deleted files from `migrate_to_v3.py`, which won't be shown with this setting alone.

I could add a similar setup to the `migrate_to_v3.py` file, but this could be annoying for downstream code that uses these functions and wants to use a different logging level / setup e.g. to a file rather than the console. 

It could work if I created the `migrate_to_v3.py` logger as an explicit child of the `cli.py` logger though i.e.
```
# in cli.py
logger = logging.getLogger("cli")

# in migrate_to_v3.py
logger = logging.getLogger("cli.migrate")
```
What do you think?

dstansby replied:
Thinking a bit, I think there are two options here:

1. Have a CLI logger and a metadata_v3.py logger (as currently implemented), and have the CLI logger configuration change functions change configuration in both logger objects.
2. Move to having a single logger for zarr-python as a whole, and just configure that appropriately when calling the CLI.

2) is a simpler option, especially when it comes to documenting to users how to configure zarr logging. It's also the option I've seen used in other projects (e.g., [matplotlib](https://matplotlib.org/stable/api/matplotlib_configuration_api.html#logging) and [sunpy](https://docs.sunpy.org/en/stable/topic_guide/logger.html)).

Can you think of any use cases to have multiple logs, as in 1) above? If there are compelling use cases then maybe we should go with multiple logging objects; otherwise, I'd be up for just going for one for the whole of zarr-python.

K-Meech replied:
Thanks @dstansby - I agree using a single logger is the simpler option. I've implemented a solution very similar to the [matplotlib setup](https://github.com/matplotlib/matplotlib/blob/2b0f3dc833aea9e9c9c3c9a31c67531c9dbdd52c/lib/matplotlib/__init__.py#L284) which seems to be working well with the CLI.

K-Meech commented on src/zarr/core/metadata/converter/cli.py:
```diff
@@ -0,0 +1,180 @@
+    remove_v2_metadata: Annotated[
+        bool,
+        typer.Option(
+            help="Remove v2 metadata (if any) from the output location, after migration is complete."
```
This is another case for writing to a separate output location - if you run e.g. `zarr migrate v3 path/to/input-zarr.zarr path/to/output-zarr.zarr`, then there will be no v2 metadata at the output location (`output-zarr.zarr`) - hence the 'if any' qualifier.

K-Meech commented:
I've made most of the requested changes to the implementation now. Any remaining I've responded to in the review thread above - @dstansby let me know if you have any suggestions for these, or for the refactored implementation / tests.

While changing the migration functions to accept `Store` directly (rather than `StoreLike`), I made some changes in `zarr/storage/_common.py`. This  refactored `make_store_path`, to add a new `make_store` function. I also updated the docstrings - the existing `make_store_path` one was a bit out of date (e.g. didn't mention `FSMap`).

Also, I noticed that I don't have any handling for consolidated metadata (`.zmetadata`) files at the moment. Would it be fine to just error if this is encountered, or should conversion be included?

dstansby commented on src/zarr/core/metadata/converter/cli.py:
```diff
@@ -0,0 +1,180 @@
+
+@app.callback()  # type: ignore[misc]
+def main(
+    verbose: Annotated[
```
Hmm, I'm not sure. Unless anyone else weighs in with a strong opinion, we can leave it as is I think. It doesn't prevent someone coming along and adding the option for `--verbose` to sub commands later.

dstansby commented on src/zarr/storage/_common.py:
```diff
@@ -339,49 +322,106 @@ async def make_store_path(
 
     if isinstance(store_like, StorePath):
         # Already a StorePath
```
```suggestion
        # Get underlying store
```

dstansby commented on tests/test_cli/test_migrate_v3.py:
```diff
@@ -0,0 +1,632 @@
+
+    zarr_array = zarr.open(local_store.root, zarr_format=3)
+    metadata = zarr_array.metadata
+    assert metadata.zarr_format == 3
+    assert metadata.node_type == "array"
+    assert metadata.shape == shape
+    assert metadata.chunk_grid == RegularChunkGrid(chunk_shape=chunks)
+    assert metadata.chunk_key_encoding == V2ChunkKeyEncoding(separator=".")
+    assert metadata.data_type == UInt16(endianness="little")
+    assert metadata.codecs == (
+        BytesCodec(endian="little"),
+        BloscCodec(typesize=2, cname="zstd", clevel=3, shuffle="shuffle", blocksize=0),
+    )
+    assert metadata.fill_value == fill_value
+    assert metadata.attributes == attributes
+    assert metadata.dimension_names is None
+    assert metadata.storage_transformers == ()
```
Instead of asserting every attribute on metadata, it might be nicer to do `assert metadata == MetadataObject(...)`. That way if multiple attributes are wrong, they will all show up in the pytest error, as opposed to with multiple asserts pytest will just stop on the first failure.

dstansby reviewed (commented):
Looking great! I think outstanding todos here are:

- Add a release note entry
- Add a new section to the [user guide docs ](https://zarr.readthedocs.io/en/stable/user-guide/index.html) to advertise the new CLI
- Work out how to do logging (see inline thread discussion)

d-v-b commented:
as a general comment about converting v2 -> v3, converting the v2 `filters, compressor` to teh v3 `codecs` is hard. I suspect a general solution will not really be possible in this PR, because it relies on features that don't exist in Zarr Python yet. I am working on this (see #3162,  https://github.com/zarr-developers/zarr-python/pull/3276), but it's a big lift. Getting PRs like #3318 merged is an important part of this process. But once we do this these features in, we can generically convert filters + compressors to codecs.

K-Meech commented:
Thanks @d-v-b - at the moment this PR:
- explicitly converts v2 `numcodecs` blosc / zstd / gzip to the v3 `zarr.codecs` equivalent.
- For any other v2 `numcodecs` codec, it tries to find a matching `numcodecs.zarr3` codec (by name), then initialises it with same settings from `.get_config`. This works fine for some codecs, but does cause issues for others e.g.  https://github.com/zarr-developers/zarr-python/issues/3256 .
 
Happy to wait on merging this PR until the other issues / PRs you mentioned are resolved.

d-v-b commented on src/zarr/metadata/migrate_v3.py:
```diff
@@ -0,0 +1,294 @@
+    return False
+
+
+def _convert_array_metadata(metadata_v2: ArrayV2Metadata) -> ArrayV3Metadata:
```
later on we can put this in `zarr.core.metadata`, e.g. in a `conversion` module

d-v-b commented:
a high-level comment about the conversion strategy. this is not a blocker for this PR, but the current conversion strategy is serial over arrays and groups. This means converting an array / group waits until the previous one is converted, and it increases the likelihood of conversion breaking in the middle, leaving a hybrid v2 / v3 hierarchy.

We could also do conversion with a concurrent API based on concurrently reading the entire input hierarchy with `dict(group.members(max_depth=None))`, preparing new metadata documents for each member, then concurrently writing those documents with [`create_hierarchy`](https://github.com/zarr-developers/zarr-python/blob/a26926cf4ec51a12ca28ff31c13ef2c8a32d51f9/src/zarr/core/sync_group.py#L51). 

This could be considered merely a performance concern, but I suspect reading and writing the hierarchy concurrently might be the difference between success and failure for this CLI when people use it on high-latency storage backends.

@K-Meech if you want to experience how your CLI tool handles IO latency, we have a [store wrapper class](https://github.com/zarr-developers/zarr-python/blob/a26926cf4ec51a12ca28ff31c13ef2c8a32d51f9/src/zarr/testing/store.py#L530) that adds latency to any other store class. Try running your conversion on a large zarr hierarchy with 100ms of latency added to `get` and `set` on a local store. If it takes more than a few seconds, then this strongly argues for moving away from the serial conversion strategy and using something concurrent.

K-Meech commented:
I've fixed the merge conflicts with this branch, and updated docstrings to reference the new `zarr.codecs.numcodecs` (rather than `numcodecs.zarr3`). I think I've addressed everyone's comments as well - @dstansby , see my replies to your comments above.
Do let me know if anything else needs changing / updating here - otherwise it might be ready to go in?

d-v-b commented:
it's going in once the tests are green. Thanks so much @K-Meech! If you don't mind, I'd like to make a post about your work over at image.sc, so people know about this functionality. This might be a good way to get some early feedback from users.

K-Meech commented:
Great - thanks @d-v-b ! I'll make some follow up issues for [additional features from the reviews on this PR](https://github.com/zarr-developers/zarr-python/pull/3257#issuecomment-3196157284).

joshmoore commented:
👏🏽 👏🏽 - The one thought I had while pondering this, @K-Meech, (and sorry if it was already discussed) is to what degree the CLI will follow the semver of the library itself. Not that anything needs doing but might be worth expectation management on the part of consumers.

d-v-b commented:
> 👏🏽 👏🏽 - The one thought I had while pondering this, @K-Meech, (and sorry if it was already discussed) is to what degree the CLI will follow the semver of the library itself. Not that anything needs doing but might be worth expectation management on the part of consumers.

the library doesn't follow semver -- we are using [effver](https://zarr.readthedocs.io/en/stable/developers/contributing.html#compatibility-and-versioning-policies)
